samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -188,98 +183,82 @@ private void validateConfig() {
-          }
-        }
-      }
-
-      // assign the partition count for intermediate streams
-      for (StreamEdge edge : joinSpecToStreamEdges.get(join)) {
-        if (edge.getPartitionCount() <= 0) {
-          log.info("Set the partition count to {} for input stream {} to the join {}.",
-              new Object[] {partitions, edge.getName(), join.getOpId()});
-          edge.setPartitionCount(partitions);
-
-          // find other joins can be inferred by setting this edge
-          for (JoinOperatorSpec op : streamEdgeToJoinSpecs.get(edge)) {
-            if (!visited.contains(op)) {
-              joinQ.add(op);
-              visited.add(op);
-            }
-          }
-        }
-      }
+  private void validateJoinInputStreamPartitions(JobGraph jobGraph) {
+    // Group joined input operator specs (input/intermediate streams) by the join
+    // operator specs corresponding to the join operations they participate in.
+    Multimap<JoinOperatorSpec, InputOperatorSpec> joinOpSpecToInputOpSpecs =
+        OperatorSpecGraphAnalyzer.getJoinedInputOperatorSpecs(jobGraph.getSpecGraph());
+
+    // Convert every group of input operator specs into a group of corresponding stream edges.
+    List<JoinedStreamsGroup> joinedStreamsGroups = new ArrayList<>();
+    for (JoinOperatorSpec joinOpSpec : joinOpSpecToInputOpSpecs.keySet()) {
+      Collection<InputOperatorSpec> joinedInputOpSpecs = joinOpSpecToInputOpSpecs.get(joinOpSpec);
+      JoinedStreamsGroup joinedStreamsGroup = getJoinedStreamsGroup(joinOpSpec.getOpId(), joinedInputOpSpecs, jobGraph);
+      joinedStreamsGroups.add(joinedStreamsGroup);
     }
+
+    /*
+     * Sort the stream edge groups by their category so they appear in this order:
+     *   1. groups composed exclusively of stream edges with set partition counts
+     *   2. groups composed of a mix of stream edges  with set/unset partition counts
+     *   3. groups composed exclusively of stream edges with unset partition counts
+     *
+     *   This guarantees that we process the most constrained stream edge groups first,
+     *   which is crucial for intermediate stream edges that are members of multiple
+     *   stream edge groups. For instance, if we have the following groups of stream
+     *   edges (partition counts in parentheses, question marks for intermediate streams):
+     *
+     *      a. e1 (16), e2 (16)
+     *      b. e2 (16), e3 (?)
+     *      c. e3 (?), e4 (?)
+     *
+     *   processing them in the above order (most constrained first) is guaranteed to
+     *   yield correct assignment of partition counts of e3 and e4 in a single scan.
+     */
+    Collections.sort(joinedStreamsGroups, Comparator.comparing(JoinedStreamsGroup::getCategory));
+
+    // Verify agreement between joined input/intermediate streams.
+    // This may involve setting partition counts of intermediate stream edges.
+    joinedStreamsGroups.forEach(ExecutionPlanner::validateJoinedStreamsGroupPartitions);
```

Comment:
Yes, validate is also assigning partition counts. It's a little difficult to separate both operations at the moment. For groups containing stream edges with known partition counts only, we just need verification — this is the easy case. On the other hand, stream edge groups with a mix of set/unset partitions require a mix of verification and assignment.

I thought of deciding whether to do verification vs assignment depending on the `JoinedStreamsGroupCategory` of a `JoinedStreamsGroup`. Problem with that is:

1. For a group with a mix of set/unset partitions, I could very easily need to do both verification and assignment within the same group, e.g. {e1 (8), e2 (?), e3 (8)}.

2. `JoinedStreamsGroupCategory` is not really reliable once we start setting partition counts (which is another reason why `StreamEdge`s are better of being immutable). For instance, by the time we process group #2 of {e1 (8), e2 (?)} and {e2 (?), e3 (8)}, it will only require verification even though its (stale) category will be `SOME_PARTITION_COUNT_SET`.

I think we can just change the verb from `validate` to something that conveys possibility of mutation. I'll try to come up with something but I'm also open to suggestions.